### PR TITLE
Upgrade action/{upload,download}-artifact to v4 [3.0]

### PR DESCRIPTION
**Pull request by levitte**

### CODE REVIEW

**Comment by nhorman** [comment]
@quarckster  downgraded these actions in 089271601a1d085f33ef7b7d8c3b6879045be370 because v4 isn't compatible with GHE, is that fixed now?

**Comment by quarckster** [comment]
> @quarckster downgraded these actions in [0892716](https://github.com/openssl/openssl/commit/089271601a1d085f33ef7b7d8c3b6879045be370) because v4 isn't compatible with GHE, is that fixed now?

I don't think so:

> actions/upload-artifact@v3 is scheduled for deprecation on November 30, 2024. [Learn more.](https://github.blog/changelog/2024-04-16-deprecation-notice-v3-of-the-artifact-actions/) Similarly, v1/v2 are scheduled for deprecation on June 30, 2024. Please update your workflow to use v4 of the artifact actions. This deprecation will not impact any existing versions of GitHub Enterprise Server being used by customers.

**Comment by levitte** [comment]
Argh.

I wonder what, exactly, is incompatible...  have we tested what happens if we try on GHE anyway?

**Comment by quarckster** [comment]
> Argh.
> 
> I wonder what, exactly, is incompatible... have we tested what happens if we try on GHE anyway?

v4 won't upload artifacts to GHE

**Comment by levitte** [comment]
Ok?  According to [github's announcement](https://github.blog/changelog/2024-04-16-deprecation-notice-v3-of-the-artifact-actions/), v3 will stop working after 5th december this year, i.e. in less than a month.

It's ok if merging these PRs is delayed until then, and I hope github will fix GHE 'til then.  Perhaps they should be notified that this is indeed an issue.

**Comment by quarckster** [comment]
> Ok? According to [github's announcement](https://github.blog/changelog/2024-04-16-deprecation-notice-v3-of-the-artifact-actions/), v3 will stop working after 5th december this year, i.e. in less than a month.
> 
> It's ok if merging these PRs is delayed until then, and I hope github will fix GHE 'til then. Perhaps they should be notified that this is indeed an issue.

That's unfortunate. In that case it would be easier to migrate our GHE mirrors to GitHub private repositories. This also removes [these template hacks](https://github.com/openssl/openssl/blob/master/.github/workflows/ci.yml#L29C14-L29C111).

**Comment by t8m** [comment]
I.e., just add `if: github.server_url == 'https://github.com'` to the upload-artifact steps.

**Comment by t8m** [comment]
> That's unfortunate. In that case it would be easier to migrate our GHE mirrors to GitHub private repositories. This also removes [these template hacks](https://github.com/openssl/openssl/blob/master/.github/workflows/ci.yml#L29C14-L29C111).

Not that I am against that in general, but we shouldn't rush this in _just because_.
IMO there is a clean and easy workaround that I proposed above.

**Comment by levitte** [comment]
> I.e., just add `if: github.server_url == 'https://github.com'` to the upload-artifact steps.

I can agree with that, at least as a temporary measure

**Comment by levitte** [comment]
> > That's unfortunate. In that case it would be easier to migrate our GHE mirrors to GitHub private repositories. This also removes [these template hacks](https://github.com/openssl/openssl/blob/master/.github/workflows/ci.yml#L29C14-L29C111).
> 
> Not that I am against that in general, but we shouldn't rush this in _just because_. IMO there is a clean and easy workaround that I proposed above.

Heh, we're already seeing a progression in that direction, and in many ways, such a move could simplify things.

**Comment by levitte** [comment]
> > I.e., just add `if: github.server_url == 'https://github.com'` to the upload-artifact steps.
> 
> I can agree with that, at least as a temporary measure

I've added changes for this in this and the other PRs

File: .github/workflows/fips-checksums.yml [hunk 1]
```diff
@@ -69,6 +69,13 @@ jobs:
       - name: save PR number
         run: echo ${{ github.event.number }} > ./artifact/pr_num
       - name: save artifact
+        if: github.server_url == 'https://github.com'
+        uses: actions/upload-artifact@v4
```
Comment (quarckster) [inline .github/workflows/fips-checksums.yml:73]:
```yaml
uses: actions/upload-artifact@${{ github.server_url == 'https://github.com' && 'v4' || 'v3' }}
```

Comment (quarckster) [reply .github/workflows/fips-checksums.yml:73]: Unfortunately, it doesn't :( https://github.com/quarckster/openssl-fork/actions/runs/11783243850/workflow#L102

Comment (levitte) [reply .github/workflows/fips-checksums.yml:73]: Yeah, I just tried a variant of the same, and came to an error on it as well.  Meh...

**Review by t8m — APPROVED** [review]
ping @openssl/committers

**Review by beldmit — APPROVED** [review]
LGTM

**Comment by levitte** [comment]
Merged

2ab98b00306877d7b57369ca9ca921ee0be37464 Upgrade action/{upload,download}-artifact to v4

**Pull request closed without merging**